### ISSUE TITLE
dynamic layout

### ISSUE THREAD

**Issue by echo66** [jwkvam/bowtie#177]
Greetings!

Is it possible to have dynamic layouts like dash? For example

```python
def layout_func(): 
  num = numpy.random.rand()
  if num > 0.5:
    return html.Div([
      html.Span('A')
    ])
  else:
    return html.Div([html.Span(i) for i in range(numpy.random.randint(100))])

app.layout = layout_func
```

I'm asking this because, in bowtie, the layouts need to be built before starting the server.

**Comment by jwkvam** [comment]
Hi @echo66 , thanks for letting me know about this feature. It's a use case I hadn't considered. As you guessed, it's not currently possible to have dynamic layouts in bowtie. The best workaround is to construct different views, but if your layouts are data dependent like in your example even that won't cut it.

The layout itself is a react component [1]. So technically it's easy to control that with Python, it just needs some socketio hooks. The tricky thing, I *think*, is how to place components in the new layout. I wonder how dash handles that. Can it have dynamic components as well as a dynamic layout?

https://github.com/jwkvam/bowtie/blob/master/bowtie/templates/view.jsx.j2

**Comment by echo66** [comment]
Dash supports dynamic components and layouts but it can't create callbacks on-the-fly, you need to create them upfront [1]. 

[1] https://community.plot.ly/t/callback-for-dynamically-created-graph/5511/4

**Comment by jwkvam** [comment]
Thanks for those examples, I wasn't able to tell if `click_2` was actually working in your last example but I understand what you mean with dynamic callbacks though.

As for implementing this in Bowtie, I'm thinking about this in 3 parts from easiest to hardest:

1. Implement dynamic layout. I'm thinking this would consist of some command I can send to views to change where components are placed and the shape of the grid. I hope this will cover most dynamic use cases and would put Bowtie in feature parity with the dash example you linked above.

2. Dynamically created components. This one's a little more tricky because the components get created at build time and the link between the Python and JS objects is established. I *think* this would require having some command to create an instance in JS and return the reference to the Python backend.

3. Dynamic callbacks. This is the one I'm least sure about because there might be some limitations with flask-socketio I'm not aware of.

This all necessarily pushes the App building phase out of the `__main__` block of code. It's not a big deal it just leaves some more global python objects lying around, but I think the trade off is worth it.

**Comment by echo66** [comment]
Hey Jacques! Thanks for the (detailed) answer.

First of all, let's explore a bit what Dash is doing. Run the following example:

```python
import dash
from dash.dependencies import Input, Output, Event, State
import dash_html_components as html
import dash_core_components as dcc

from pprint import pprint

import time


counter = {"c": 2}

app = dash.Dash()
app.config["suppress_callback_exceptions"] = True

print("App start")


def serve_layout(*args):
    print("Calling layout")
    pprint(args)
    return html.Div(id="master", children=[
        dcc.Interval(id="interval", interval=500),
        html.Div(id="main", children=[
            html.Div(id="output-1"),
            html.Button(id="b1", children="button 1"),
            html.Div(id="out-0"),
            html.Div(id="out-1"),
            html.Button(id="b3", children="button 3")
        ])
    ])


app.layout = serve_layout


@app.callback(
    output=Output("output-1", "children"),
    events=[Event("b1", "click")],
    state=[State("output-1", "children")]
)
def click(children):
    c = counter["c"]
    counter["c"] += 1
    #print(c)
    pprint(children)
    #pprint(app.callback_map)

    @app.callback(
        output=Output("output-{}".format(c), "children"),
        events=[Event("b{}".format(c), "click")],
    )
    def click_2():
        return "b2 clicked"

    if children:
        children['props']['children'][1]['props']['children'] = 'oioi'
    else:
        children = html.Div([
            html.Div(id="output-{}".format(c)),
            html.Button(id="b{}".format(c),
                        children="button {}".format(c))
        ])

    print(children)

    return children


@app.callback(
    output=Output("out-1", "children"),
    events=[Event("b3", "click")]
)
def g():
    time.sleep(2)
    c = counter["c"]
    counter["c"] += 1
    return c


@app.callback(
    output=Output("out-0", "children"),
    events=[Event("interval", "interval")]
)
def f():
    c = counter["c"]
    counter["c"] += 1
    return c


if __name__ == "__main__":
    app.run_server(debug=False, threaded=True)
```

Start the code above and press "button 1" once and you get this on the console: 

```
Div([Div(id='output-2'), Button(children='button 2', id='b2')])
```

Press it the second time and you get:

```json
{"namespace": "dash_html_components",
 "props": {"children": [{"namespace": "dash_html_components",
                         "props": {"children": None, "id": "output-5"},
                         "type": "Div"},
                        {"namespace": "dash_html_components",
                         "props": {"children": "oioi", "id": "b5"},
                         "type": "Button"}]},
 "type": "Div"}
```

Although the results seem different, the first result will be converted (internally) to the form of the second result and it will be sent to the client, resulting in a update of the virtual DOM (by Dash front-end part). After that, this is what I think that happens:

(1) Dash front-end stores the layout hierarchy in a (redux) state field. 
(2) Every time the state changes, Dash front-end will go through the layout data recursively (https://github.com/plotly/dash-renderer/blob/master/src/TreeContainer.js#L48) and render the react components.


The main difference between bowtie and dash layout rendering approaches is that bowtie creates the entire layout in "compilation time".

**Comment by jwkvam** [comment]
Thanks again for helping me understand. That looks quite powerful, I hope that we can find a solution in bowtie that's just as powerful and hopefully simpler. I still find all the html a bit hard to read myself so I've tended to remove that from bowtie's api.

Let me try to sketch out how I envision this working with bowtie with some psuedo-code. I'll have 3 buttons, but only 2 are visible at first. Clicking on one will make the 3rd one visible and modify the layout accordingly. As you likely know, Bowtie uses css grid to layout all the components.

```
from bowtie import App, command
from bowtie.control import Button
buttons = [Button(label=f'button {i}') for i in range(3)]

app = App(rows=2)
app.add(buttons[0])
app.add(buttons[1])

# this will keep inserting the same component into the first row.
# if this gets clicked 5 times we'll have 5+2 rows on display
# the first 5 rows will display the same button
@app.subscribe(buttons[0].on_click)
def add_last_button():
    # add another css grid row at the top
    app.do_row(index=0)
    # put buttons[2] into the new cell location
    app.do_add(buttons[2], row=0)

@command
def run():
    app.build()
```

Out of curiosity, you know Dash quite deeply (much more than me) and have invested some time learning it. Is there something you find appealing about Bowtie? Otherwise I assume you would continue using Dash?

**Comment by echo66** [comment]
As far as I know, Bowtie uses web sockets. And, as such, it supports events triggered by the server, unlike Dash. With Dash, the only way to have live updates is to query the server with something like [dash_core_components.Interval](https://github.com/plotly/dash-core-components/blob/master/src/components/Interval.react.js).

By the way, if you need to build the application layout before serving it, how can you add those buttons in runtime? Doesn't Bowtie generate, during the build process, a set of React files based on templates?

**Comment by jwkvam** [comment]
You're correct, with Bowtie, the server can trigger updates and broadcast them to all clients. That's why I can have timers running on the server. That's interesting since I suppose that could help you push needing to scale until later.

You're also correct on your second point. That call would have no effect when serving the app. I think for clarification. Those calls should be put in the `run` function.
